gsdk/src/metadata/errors.rs

```diff
@@ -0,0 +1,124 @@
+
+use std::{
+    error::Error,
+    fmt::{Debug, Display},
+};
+use subxt::ext::codec::Decode;
+
+// TODO: refactor this after #2618
+macro_rules! export_module_error {
+    ($($path:ident)::* => $error:ident => $index:expr) => {
+        pub use crate::metadata::runtime_types::$($path)::*::pallet::Error as $error;
+    };
+    ($($($path:ident)::* => $error:ident => $index:expr),*) => {
+        $(
+            export_module_error!($($path)::* => $error => $index);
+        )*
+
+        /// Common error type for runtime modules.
+        #[derive(Debug, PartialEq, Eq)]
+        pub enum ModuleError {
+            $($error($error)),*,
+            Unknown {
+                pallet_index: u8,
+                error: [u8; 4],
+            }
+        }
+
+        impl Display for ModuleError {
+             fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                 write!(f, "{:?}", self)
+             }
+        }
+
+        impl std::error::Error for ModuleError {}
+
+        impl From<subxt::error::ModuleError> for ModuleError {
+            fn from(error: subxt::error::ModuleError) -> ModuleError {
+                match error.error_data.pallet_index {
+                     $($index => match $error::decode(&mut [error.error_data.error[0]].as_ref()) {
+                         Ok(e) => ModuleError::$error(e),
+                         Err(_) => ModuleError::Unknown {
+                             pallet_index: error.error_data.pallet_index,
+                             error: error.error_data.error,
+                         },
+                     }),*,
+                     // `pallet_fellowship_referenda => 19`
+                     //
+                     // shares the same error with
+                     //
+                     // `pallet_fellowship_collective => 18`
+                     19 => {
+                         let mb_error = RanckedCollective::decode(&mut error.error_data.error.as_ref());
+                         match mb_error {
+                             Ok(e) => ModuleError::RanckedCollective(e),
+                             Err(_) => ModuleError::Unknown {
+                                 pallet_index: error.error_data.pallet_index,
+                                 error: error.error_data.error,
+                             },
+                         }
+                     },
+                     _ => ModuleError::Unknown {
+                         pallet_index: error.error_data.pallet_index,
+                         error: error.error_data.error,
+                     }
+                }
+            }
+        }
+    };
+}
+
+// Re-exports module errors from runtime types.
+//
+//
+// # NOTE
+//
+// pallets that don't have `Error` type.
+//
+// - pallet_transaction_payment
+// - pallet_airdrop
+//
+// pallets that share the same `errors::RankedCollective`
+//
+// - pallet_fellowship_collective => 18
+// - pallet_fellowship_referenda => 19
+export_module_error! {
```

Comment:
Can we somehow go with discriminant safety?
`construct_runtime` doesn't support anything except literals, but may there by some workaround?